### PR TITLE
Add email subject when sharing with mail apps

### PR DESCRIPTION
**Pull request by dominik-weber**
Include an email subject when sharing with mail apps on Android and iOS. This was already working on UWP, so these changes should make the plugin more consistent across platforms.

Please ignore the changed whitespaces, I have the Format on Save extension installed ;)

### CODE REVIEW

File: Share/Share.Plugin.Android/ShareImplementation.cs [hunk 1]
```diff
@@ -66,45 +68,54 @@ public class ShareImplementation : IShare
...
+
             var intent = new Intent(Intent.ActionSend);
             intent.SetType("text/plain");
-            intent.PutExtra(Intent.ExtraText, url ?? string.Empty);
-            intent.PutExtra(Intent.ExtraSubject, message ?? string.Empty);
+            intent.PutExtra(Intent.ExtraText, string.Join(Environment.NewLine, items));
+            if (title != null)
+                intent.PutExtra(Intent.ExtraSubject, title);
 
-            var chooserIntent = Intent.CreateChooser(intent, title ?? string.Empty);
```
Comment (jamesmontemagno) [inline Share/Share.Plugin.Android/ShareImplementation.cs]: any reason to remove this? What if it is null?

File: Share/Share.Plugin.Android/ShareImplementation.cs [hunk 2]
```diff
@@ -66,45 +68,54 @@ public class ShareImplementation : IShare
...
+                items.Add(message);
+            if (url != null)
+                items.Add(url);
+
             var intent = new Intent(Intent.ActionSend);
             intent.SetType("text/plain");
-            intent.PutExtra(Intent.ExtraText, url ?? string.Empty);
-            intent.PutExtra(Intent.ExtraSubject, message ?? string.Empty);
+            intent.PutExtra(Intent.ExtraText, string.Join(Environment.NewLine, items));
+            if (title != null)
```
Comment (jamesmontemagno) [inline Share/Share.Plugin.Android/ShareImplementation.cs]: Should we check against null or whitespace?

File: Share/Share.Plugin.Android/ShareImplementation.cs (hunk 1 above)
Comment (jamesmontemagno) [inline Share/Share.Plugin.Android/ShareImplementation.cs]: any reason to remove this ?? check? I assume null will throw an exception here.

File: Share/Share.Plugin.Android/ShareImplementation.cs (hunk 2 above)
Comment (jamesmontemagno) [inline Share/Share.Plugin.Android/ShareImplementation.cs]: Should we check against string.IsNullorWhitespace?

File: Share/Share.Plugin.Android/ShareImplementation.cs (hunk 1 above)
Comment (dominik-weber) [inline Share/Share.Plugin.Android/ShareImplementation.cs]: Sorry, totally forgot about this one. Specifying string.Empty results in a chooser without title, specifying null creates a chooser with a system default title. I think this way we can let the developer decide if he wants to have the default title (null) or no title (string.Empty)

https://developer.android.com/reference/android/content/Intent.html#createChooser(android.content.Intent, java.lang.CharSequence)

> title	CharSequence: Optional title that will be displayed in the chooser.

File: Share/Share.Plugin.Android/ShareImplementation.cs (hunk 2 above)
Comment (dominik-weber) [inline Share/Share.Plugin.Android/ShareImplementation.cs]: Hm... we could, however I believe null and string.Empty will not make a difference (setting the email subject to string.Empty is basically a no-op) and I'm not a huge fan of whitespace checks

**Comment by jamesmontemagno** [comment]
in the current beta, let me know how it goes.

**Comment by dominik-weber** [comment]
works perfectly, thanks!
Before you release this version in the stable channel, I was considering implementing the following changes:

    Task Share(ShareData data, ShareOptions options = null)

To be more consistent with the OpenBrowser method. Also we could implement platform agnostic properties like message, url and title in the ShareData object and platform specific values in the ShareOptions object:

    public class ShareData
    {
        public string Title { get; set; }
        public string Message { get; set; }
        public string Url { get; set; }
    }

    public class ShareOptions
    {
        public string ChooserTitle { get; set; } // only android
        public string[] ExcludedUIActivityTypes { get; set; } // only ios
    }

By adding a new method (and marking the old ones as deprecated) we can also hint the user to the behavioral changes.

What do you think about this?

**Comment by jamesmontemagno** [comment]
Sure, maybe make it a new method so we don't break apps?

Best,
James

**Comment by dominik-weber** [comment]
I'm currently working on the additional changes, however I'm wondering why you catch all exceptions in ShareImplementation.Share and OpenBrowser on iOS and Windows? Doesn't this violate the fail fast principle? How would the developer know wether the call was successful?

**Comment by jamesmontemagno** [comment]
That is a good question, it could return a bool for success or failure. I just seem to try/catch everything else a lot of consumers don't and then they may pass it something that doesn't work. I am mostly just over worried

**Comment by dominik-weber** [comment]
Hmm, should I change it to return a bool then?

**Comment by jamesmontemagno** [comment]
sure. I think it is just Task so changing to Task<bool> wont be breaking at all.

**Comment by dominik-weber** [comment]
done: https://github.com/jguertl/SharePlugin/pull/32

**Pull request merged**